### PR TITLE
Remove the -Zinsert-sideeffect

### PR DESCRIPTION
**Pull request by nagisa**
This removes all of the code we had in place to work-around LLVM's
handling of forward progress. From this removal excluded is a workaround
where we'd insert a `sideeffect` into clearly infinite loops such as
`loop {}`. This code remains conditionally effective when the LLVM
version is earlier than 12.0, which fixed the forward progress related
miscompilations at their root.

### CODE REVIEW

**Comment by nagisa** [comment]
cc @nikic @Mark-Simulacrum

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 1]
```diff
@@ -334,8 +334,11 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
         self.call(expect, &[cond, self.const_bool(expected)], None)
     }
 
-    fn sideeffect(&mut self, unconditional: bool) {
-        if unconditional || self.tcx.sess.opts.debugging_opts.insert_sideeffect {
+    fn sideeffect(&mut self) {
+        // This kind of check would make a ton of sense in the caller, but currently the only
+        // caller of this function is in `rustc_codegen_ssa`, which is agnostic to whether LLVM
+        // codegen backend being used, and so is unable to check the LLVM version.
+        if llvm_util::get_version() < (12, 0, 0) {
```
Comment (Mark-Simulacrum) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:341]: This should be called pretty rarely I think, but I am a bit worried about the performance - get_version makes 3 separate FFI calls after all. We'll see what perf says though.

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

I would expect either improved or unchanged performance.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs (hunk 1 above)
Comment (nagisa) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:341]: Yeah, we can just call the LLVMRustGetMajorVersion directly here.

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 2]
```diff
@@ -963,22 +927,16 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
...
+                    // (e.g., in embedded code).
                     //
-                    // The `true` here means we insert side effects regardless
-                    // of -Zinsert-sideeffect being passed on unconditional
-                    // branching to the same basic block.
-                    bx.sideeffect(true);
-                } else {
-                    helper.maybe_sideeffect(self.mir, &mut bx, &[target]);
+                    // NB: the `sideeffect` currently checks for the LLVM version used internally.
+                    bx.sideeffect();
```
Comment (estebank) [inline compiler/rustc_codegen_ssa/src/mir/block.rs:937]: So this is the _only_ call to `sideeffect` left in the codebase? Won't this cause a miscompilation for the removed cases when run under LLVM <12?

Comment (nagisa) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:937]: All the other calls did not generate a `sideeffect` unless an unstable flag was specified.

**Pull request merged**